src/CommandLine.cc

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "CommandLine.h"
+#include "sbuf/SBuf.h"
+
+static void
+ResetGetopt(const bool allowStderrWarnings)
+{
+    opterr = allowStderrWarnings;
+    // Resetting optind to zero instead of conventional '1' has an
+    // advantage, since it also resets getopt(3) global state.
+    // getopt(3) always skips argv[0], even if optind is zero
+    optind = 0;
+}
+
+CommandLine::CommandLine(int argC, char *argV[], const char &shortRules, const LongOption *longRules):
+    argv_(),
+    shortOptions_(xstrdup(&shortRules)),
+    longOptions_()
+{
+    assert(argC > 0); // C++ main() requirement that makes our arg0() safe
+
+    /* copy argV items */
+    argv_.reserve(argC+1);
+    for (int i = 0; i < argC; ++i)
+        argv_.push_back(xstrdup(argV[i]));
+    argv_.push_back(nullptr); // POSIX argv "must be terminated by a null pointer"
+
+    /* copy grammar rules for the long options */
+    if (longRules) {
+        for (auto longOption = longRules; longOption->name; ++longOption)
+            longOptions_.push_back(Option(*longOption));
```

Comment:
please prefer emplace over push here. the earlier ones could do with it too for consistency, but do not gain efficiency from move semantics like these do.

Reply:
Done.